File: backend/app/games/codenames/clue_validation.py
"""Shared Codenames clue validation for human and AI spymasters."""

MIN_SUBSTRING_CLUE_LEN = 3


def normalize_clue_word(clue: str) -> str:
    return clue.strip().upper()


def board_words_upper(cards: list[dict]) -> set[str]:
    return {c["word"].upper() for c in cards}


def conflicting_board_word(clue: str, board_words: set[str]) -> str | None:
    """Return the board word that conflicts with the clue, or None."""
    clue = normalize_clue_word(clue)
    if clue in board_words:
        return clue
    if len(clue) >= MIN_SUBSTRING_CLUE_LEN:
        for word in board_words:
            if clue in word or word in clue:
                return word
    return None


def validate_clue_word(clue: str, board_words: set[str]) -> str | None:
    """Return an error message if the clue word is invalid, else None."""
    normalized = normalize_clue_word(clue)
    if not normalized:
        return "Clue must be a single word"
    if len(normalized.split()) > 1:
        return "Clue must be a single word"

    conflict = conflicting_board_word(normalized, board_words)
    if conflict:
        if normalized == conflict:
            return "Clue cannot match a word on the board"
        return f"Clue is too similar to board word '{conflict.title()}'"
    return None

File: frontend/src/games/codenames/clueValidation.ts
const MIN_SUBSTRING_CLUE_LEN = 3

export function normalizeClueWord(clue: string): string {
  return clue.trim().toUpperCase()
}

export function getBoardWords(cards: { word: string }[] | string[]): Set<string> {
  const words =
    cards.length === 0 || typeof cards[0] === 'string'
      ? (cards as string[])
      : (cards as { word: string }[]).map((c) => c.word)
  return new Set(words.map((word) => word.toUpperCase()))
}

export function conflictingBoardWord(clue: string, boardWords: Set<string>): string | null {
  const normalized = normalizeClueWord(clue)
  if (boardWords.has(normalized)) {
    return normalized
  }
  if (normalized.length >= MIN_SUBSTRING_CLUE_LEN) {
    for (const word of boardWords) {
      if (normalized.includes(word) || word.includes(normalized)) {
        return word
      }
    }
  }
  return null
}

export function validateClueWord(clue: string, boardWords: Set<string>): string | null {
  const normalized = normalizeClueWord(clue)
  if (!normalized) {
    return 'Clue must be a single word'
  }
  if (normalized.split(/\s+/).length > 1) {
    return 'Clue must be a single word'
  }

  const conflict = conflictingBoardWord(normalized, boardWords)
  if (conflict) {
    if (normalized === conflict) {
      return 'Clue cannot match a word on the board'
    }
    return `Clue is too similar to board word '${conflict.charAt(0)}${conflict.slice(1).toLowerCase()}'`
  }
  return null
}
